src/test/groovy/graphql/execution/ValuesResolverE2ETest.groovy

```diff
@@ -0,0 +1,136 @@
+                                                        .name("limit")
+                                                        .type(Scalars.GraphQLInt))
+                                                .field(offset -> offset
+                                                        .name("offset")
+                                                        .type(Scalars.GraphQLInt))
+                                                .build())))
+                        .build())
+                .codeRegistry(GraphQLCodeRegistry.newCodeRegistry()
+                        .dataFetcher(FieldCoordinates.coordinates("Query", "items"), dataFetcher)
+                        .build())
+                .build()
+
+        GraphQL gql = GraphQL.newGraphQL(schema).build()
+
+        Map<String, Object> vars = new HashMap<>()
+        vars.put("limit", 5)
+        vars.put("offset", 0)
+
+        ExecutionInput ei = ExecutionInput.newExecutionInput()
+                .query("query Items( \$limit: Int, \$offset: Int) {\n" +
+                        "  items(\n" +
+                        "    pagination: {limit: \$limit, offset: \$offset} \n" +
+                        "  )\n" +
+                        "}")
+                .variables(vars)
+                .build()
+
+        when:
+        ExecutionResult result = gql.execute( ei)
+        then:
+        result.errors.isEmpty()
+        result.data == [items : ["limit=5", "offset=0"]]
+    }
+
+    def "issue 3276 - should end up in validation errors because location defaults are not present"() {
+        def sdl = '''
+            type Query {
+                items(pagination: Pagination = {limit: 1, offset: 1}): [String]
+            }
+            input Pagination {
+                limit: Int! #non-null this time, no default value
+                offset: Int! #non-null this time, no default value
+            }
+        '''
+        DataFetcher df = { DataFetchingEnvironment env ->
+            def pagination = env.getArgument("pagination") as Map<String, Integer>
+            def strings = pagination.entrySet().collect { entry -> entry.key + "=" + entry.value }
+            return strings
+        }
+        def schema = TestUtil.schema(sdl, [Query: [items: df]])
+        def graphQL = GraphQL.newGraphQL(schema).build()
+
+        when:
+        def ei = ExecutionInput.newExecutionInput('''
+            query Items( $limit: Int, $offset: Int) {
+                 items(
+                    pagination: {limit: $limit, offset: $offset} 
+                )
+            }
+            ''').variables([limit: 5, offset: null]).build()
+        def er = graphQL.execute(ei)
+        then:
+        er.errors.size() == 2
+        er.errors[0].message == "Validation error (VariableTypeMismatch@[items]) : Variable type 'Int' does not match expected type 'Int!'"
+        er.errors[1].message == "Validation error (VariableTypeMismatch@[items]) : Variable type 'Int' does not match expected type 'Int!'"
+    }
+}
```

Comment:
These cases are covered in unit tests BUT I wanted to have an end to end test to ensure we didnt miss something